### ISSUE TITLE
Invalid xAPI statements (coming from Interactive video)

### ISSUE THREAD

**Issue by nadavkav** [h5p/h5p-moodle-plugin#172]
Hello H5P team,
Using latest hvp version 1.5 with latest interactive video libraries, I am getting "Invalid xAPI" errors from events that are generated by the interactive video package and send as xAPI statements to mod/hvp. (which, as result, not saved into the Moodle mdl_hvp_xapi_results)

Temporarily commenting out the xapi validation code: 
https://github.com/h5p/h5p-moodle-plugin/blob/master/classes/xapi_result.php#L67
Seems to send and save the statements to the mdl_hvp_xapi_results without any other issues. and they are properly shown on the reports page. weird.
I have not yet tried to debug and find out what exactly is annoying the validation functions.
So just letting you know, and I will keep you updated when I find the reason.

**Comment by thomasmars** [comment]
Hi,
I'm not able to reproduce this. 
What version is your Interactive Video ? 
Have you modified Interactive Video in any way ? 
What interactions do you have within your Interactive Video ?

Thanks for reporting this, I hope we can figure it out swiftly

**Comment by nadavkav** [comment]
I am using v1.5 of Moodle mod/mvp with latest H5P Interactive video from h5p.org website.
Looking at the code (xdebug), I am not sure how it works for you.
it seems you are looking for $this->getObjectDefinition('interactionType') on null objects coming off the xapi statement coming off the Interactive video.
![snapshot203](https://user-images.githubusercontent.com/231168/31319637-da4fa62a-ac6f-11e7-9053-2f942d40f463.png)
How about removing null $child (s) from $children ?
https://github.com/h5p/h5p-php-report/blob/78233cf2810e6053d38de90e92b7a1b0f2b26e19/h5p-report-xapi-data.class.php#L47

**Comment by nadavkav** [comment]
I think, in the case of Interactive video and also Presentation, that each "child" interaction (like a fill-in, choice, drag-drop, ...) should send xAPI statements to Moodle on their own, and not wait until the entire activity finished, and then send a full multi part (compound) statement. What do you think?

**Comment by thomasmars** [comment]
Thank you for the clarification, I was able to reproduce it now. If you have a child that returns null then the validation fails, this can be reproduced with any static elements inside IV, such as "text" or "label".
I think any child that returns null should be skipped in the validation, so that it will go through, since these elements should not be validated. I think the best way of doing this is to skip ```$child === null``` occurences in getChildren() inside h5p-report-xapi-data.class.php.

I have created an issue in the H5P issue tracker: https://h5ptechnology.atlassian.net/browse/HFP-1629
Feel free to have a go at it :)

Regarding your second comment, all "child" interactions does already send their own xAPI statements. The final full compound statement is only extracted when the main interaction (in this case IV) sends a "completed" or  "answered" event. Sending all xAPI info at the same time is a convenience that the [getXAPIData()](https://h5p.org/documentation/developers/contracts#guides-header-6) contract provides, which makes us able to render the complete report from one statement, instead of storing all child-parent statements in a tree and finally combining them at the end.
